perf(transformer/styled-components): cache block name

Dunqing opened the pull request:
Caching block name, which is obtained from the file stem or directory name, because it is always the same in every component.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12200?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12200** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12200?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12200?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12199** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12199?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12197** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12197?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12200).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -680,13 +683,15 @@ impl<'a> StyledComponents<'a, '_> {
-    /// Returns the display name which infers the component name or gets from the file name.
-    fn get_display_name(&self, ctx: &TraverseCtx<'a>) -> Option<&'a str> {
-        let component_name = Self::get_component_name(ctx);
+    /// Returns the block name based on the file stem or parent directory name.
+    fn get_block_name(&mut self, ctx: &TraverseCtx<'a>) -> Option<&'a str> {
```
Block name is never inserted into AST. I think it should be a `String`, rather than allocating it into arena.

Dunqing replied:
Done!

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -701,6 +706,15 @@ impl<'a> StyledComponents<'a, '_> {
                 file_stem
             };
 
+            ctx.ast.allocator.alloc_str(block_name)
```
There's an unnecessary allocation in `to_string` in the block above:

```diff
- let block_name = if self.options.meaningless_file_names.contains(&file_stem.to_string())
+ let block_name = if self.options.meaningless_file_names.iter().any(|file_name| file_stem == file_name)
```

Dunqing replied:
Yes, `meaningless_file_names` has a few items always, so this is fine to use `iter.any`

Dunqing replied:
I just realized that 😅

overlookmotel reviewed (commented):
Nice! A couple of comments below.

Dunqing commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -680,27 +683,38 @@ impl<'a> StyledComponents<'a, '_> {
+    }
+
+    /// Returns the display name which infers the component name or gets from the file name.
+    fn get_display_name(&mut self, ctx: &TraverseCtx<'a>) -> Option<&'a str> {
+        let component_name = Self::get_component_name(ctx);
 
+        if let Some(block_name) = self.get_block_name() {
             if let Some(component_name) = component_name {
                 if block_name == component_name {
                     return Some(ctx.ast.str(component_name.as_str()));
```
Good catch

Dunqing replied:
Wait, this is totally wrong. If the lifetime is incorrect, the compiler will complain about it.

Dunqing committed:
perf(transformer/styled-components): cache block name (#12200)

Caching block name, which is obtained from the file stem or directory name, because it is always the same in every component.
